Mage/src/main/java/mage/abilities/effects/common/BalanceEffect.java

```diff
@@ -108,34 +86,41 @@ public boolean apply(Game game, Ability source) {
             if (player == null) {
                 continue;
             }
-            TargetControlledPermanent target = new TargetControlledPermanent(minCreature, minCreature, filter2, true);
-            if (!target.choose(Outcome.Sacrifice, player.getId(), source.getSourceId(), game)) {
+            Cards cards = new CardsImpl();
+            TargetCardInHand target = new TargetCardInHand(minCard, filter3);
+            if (!target.choose(Outcome.Discard, player.getId(), source.getSourceId(), game)) {
                 continue;
             }
-            for (Permanent permanent : game.getBattlefield().getActivePermanents(StaticFilters.FILTER_CONTROLLED_CREATURE, player.getId(), source.getSourceId(), game)) {
-                if (permanent != null && !target.getTargets().contains(permanent.getId())) {
-                    creaturesToSacrifice.add(permanent);
+            for (Card card : player.getHand().getCards(game)) {
+                if (card != null && !target.getTargets().contains(card.getId())) {
+                    cards.add(card);
                 }
             }
+            cardsToDiscard.put(playerId, cards);
         }
 
-        for (UUID cardId : creaturesToSacrifice) {
-            Permanent permanent = game.getPermanent(cardId);
-            if (permanent != null) {
-                permanent.sacrifice(source.getSourceId(), game);
+        for (UUID playerId : game.getState().getPlayersInRange(controller.getId(), game)) {
+            Player player = game.getPlayer(playerId);
+            if (player != null && cardsToDiscard.get(playerId) != null) {
+                player.discard(cardsToDiscard.get(playerId), source, game);
             }
         }
 
-        //Cards in hand
-        int minCard = Integer.MAX_VALUE;
-        Map<UUID, Cards> cardsToDiscard = new HashMap<>(2);
+        return true;
+    }
+
+    private Cards getPermanentsToSacrifice(Game game, Ability source, Player controller,
+                                           FilterControlledPermanent filterPermanent, FilterControlledPermanent filterPermanentDialog) {
+        int min = Integer.MAX_VALUE;
+        Cards permToSacrifice = new CardsImpl();
```

Comment:
`Cards` is hashset, so sacrifice order and logs must be random. The best way -- use simple list -- then logs will be "sorted" by choosed players (more user friendly).

Hmm. There are oracle text: Each player will know the choices made by the players who chose before them. So you must:
1. Collect chooses to keep/sacrifice lands (inform other players about chooses after each cycle);
2. Sacrifice;
3. Repeat 1-2 for creatures;
4. Repeat 1-2 for hands.

Full oracle text:

First the player whose turn it is chooses which lands (if any) to keep, then each other player in turn order does the same. Each player will know the choices made by the players who chose before them. All of the unchosen lands are then sacrificed simultaneously. Then the process is repeated for cards in hand, except that no cards are revealed until all players have chosen what to discard, at which point those cards are all discarded simultaneously. Lastly, the process is repeated for creatures, and players will again know earlier choices made when deciding what to sacrifice. All of the unchosen creatures are then sacrificed simultaneously. (2016-06-08)